go-controller/pkg/allocator/pod/pod_annotation_test.go

```diff
@@ -880,6 +884,176 @@ func Test_allocatePodAnnotationWithRollback(t *testing.T) {
+					nextIPs:          ovntest.MustParseIPNets("192.168.0.3/24"),
+					allocateIPsError: ipam.ErrAllocated, // Should be skipped because network already allocated
+				},
+			},
+			wantPodAnnotation: &util.PodAnnotation{
+				IPs: ovntest.MustParseIPNets("192.168.0.150/24"),
+				MAC: util.IPAddrToHWAddr(ovntest.MustParseIPNets("192.168.0.150/24")[0].IP),
+			},
+			// No wantUpdatedPod because annotation already exists and no changes needed
+		},
+		{
+			// Test ErrAllocated is not skipped when EnablePreconfiguredUDNAddresses enabled and no special conditions met
+			name:                            "ErrAllocated should not be skipped when EnablePreconfiguredUDNAddresses enabled and no skip conditions met",
+			ipam:                            true,
+			enablePreconfiguredUDNAddresses: true,
+			args: args{
+				network: &nadapi.NetworkSelectionElement{
+					IPRequest: []string{"192.168.0.100/24"}, // Request specific IP to trigger AllocateIPs call
+				},
+				reallocate: false, // Don't reallocate on error
+				ipAllocator: &ipAllocatorStub{
+					nextIPs:          ovntest.MustParseIPNets("192.168.0.3/24"),
+					allocateIPsError: ipam.ErrAllocated,
+				},
+			},
+			wantErr: true,
+		},
+		{
+			// Test ErrAllocated is always skipped with EnablePreconfiguredUDNAddresses disabled (legacy behavior)
+			name:                            "ErrAllocated should be skipped when EnablePreconfiguredUDNAddresses disabled",
+			ipam:                            true,
+			persistentIPAllocation:          true,
+			enablePreconfiguredUDNAddresses: false,
+			args: args{
+				network: &nadapi.NetworkSelectionElement{
+					IPAMClaimReference: "my-ipam-claim",
+				},
+				ipamClaim: &ipamclaimsapi.IPAMClaim{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "my-ipam-claim",
+					},
+					Status: ipamclaimsapi.IPAMClaimStatus{
+						IPs: []string{}, // No IPs allocated, but should still be skipped
+					},
+				},
+				ipAllocator: &ipAllocatorStub{
+					nextIPs:          ovntest.MustParseIPNets("192.168.0.200/24"),
+					allocateIPsError: ipam.ErrAllocated,
+				},
+			},
+			wantUpdatedPod: true,
+			wantPodAnnotation: &util.PodAnnotation{
+				IPs:      ovntest.MustParseIPNets("192.168.0.200/24"),
+				MAC:      util.IPAddrToHWAddr(ovntest.MustParseIPNets("192.168.0.200/24")[0].IP),
+				Gateways: []net.IP{ovntest.MustParseIP("192.168.0.1").To4()},
+				Routes: []util.PodRoute{
+					{
+						Dest: &net.IPNet{
+							IP:   ovntest.MustParseIP("100.65.0.0").To4(),
+							Mask: net.CIDRMask(16, 32),
+						},
+						NextHop: ovntest.MustParseIP("192.168.0.1").To4(),
+					},
+				},
+				Role: types.NetworkRolePrimary,
+			},
+			wantReleasedIPsOnRollback: ovntest.MustParseIPNets("192.168.0.200/24"), // IPs should be allocated and released on rollback
+			role:                      types.NetworkRolePrimary,
+		},
+		{
+			// Test ErrAllocated with network annotation already persisted and no ipamClaim (regular pod) - should skip ErrAllocated
```

Comment:
what's the difference from https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5411/files#diff-883a6c6f6643721aa0f69b0d8f2c8ddea131441391492c8634a3376b9b70267bR888 ?